File: src/core/state-anchor.ts
import { compactPrompt } from "./tokens.js";

export interface StateAnchor {
  goal: string;
  done: string;
  next: string;
}

export function buildStateAnchor(rawPrompt: string, optimized: string, previous?: StateAnchor): StateAnchor {
  const goal = previous?.goal ?? `Fix ${compactPrompt(rawPrompt).replace(/^Fix\s+/i, "").trim()}`;
  const done = summarizeDone(optimized) ?? previous?.done ?? "Pending";
  const next = previous?.next ?? inferNext(optimized);
  return { goal, done, next };
}

export function formatStateAnchor(anchor: StateAnchor): string {
  return [
    `[GOAL] ${anchor.goal}`,
    `[DONE] ${anchor.done}`,
    `[NEXT] ${anchor.next}`
  ].join("\n");
}

function summarizeDone(optimized: string): string | null {
  const match = optimized.match(/File:\s*([^|]+)/i);
  if (!match) {
    return null;
  }
  return `Prepared patch scope for ${match[1].trim()}`;
}

function inferNext(optimized: string): string {
  if (/Error:/i.test(optimized)) {
    return "Patch failing path and verify tests";
  }
  return "Apply minimal diff and validate constraints";
}

File: core/state_anchor.py
"""Three-line state anchor for low-token continuity."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(slots=True)
class StateAnchor:
    goal: str
    done: str
    next: str

    def to_prompt(self) -> str:
        return f"[GOAL] {self.goal}\n[DONE] {self.done}\n[NEXT] {self.next}"


def build_state_anchor(raw_prompt: str, optimized_prompt: str, previous: StateAnchor | None = None) -> StateAnchor:
    if previous:
        goal = previous.goal
        done = previous.done
        next_step = previous.next
    else:
        goal = f"Fix {raw_prompt.strip()}" if raw_prompt.strip() else "Fix pending task"
        done = "Pending"
        next_step = "Apply minimal diff and run tests"

    if "File:" in optimized_prompt:
        done = "Prepared scoped patch instructions"
    if "Error:" in optimized_prompt:
        next_step = "Patch failing path and re-run checks"

    return StateAnchor(goal=goal, done=done, next=next_step)

